src/ReadWriteLocks.jl

```diff
@@ -1,42 +1,39 @@
 module ReadWriteLocks
 
-export ReadWriteLock, read_lock, write_lock, lock!, unlock!
-
-if isdefined(Base, :lock!)
-    import Base: lock!
-end
-
-if isdefined(Base, :unlock!)
-    import Base: unlock!
+using Base: lock, unlock
+if VERSION < v"1.2.0-DEV.28"
+    using Base.Threads: AbstractLock
+else
+    using Base: AbstractLock
 end
 
-if !isdefined(Base, :Mutex)
-    typealias Mutex ReentrantLock
-
-    lock!(x::Mutex) = lock(x)
-    unlock!(x::Mutex) = unlock(x)
-end
-
-abstract _Lock
+export ReadWriteLock, read_lock, write_lock, lock!, unlock!
 
-immutable ReadLock{T<:_Lock}
+struct ReadLock{T<:AbstractLock}
     rwlock::T
 end
 
-immutable WriteLock{T<:_Lock}
+struct WriteLock{T<:AbstractLock}
     rwlock::T
 end
 
-type ReadWriteLock <: _Lock
+# Need Julia VERSION > v"1.2.0-DEV.28` to have `ReentrantLock <: AbstractLock`
+LockTypes = Union{AbstractLock, ReentrantLock}
```

Comment:
should this also include `Base.AlwaysLockedST`?

Reply:
No, I don't think that's intended for public use